### PR TITLE
make item count background work

### PR DESCRIPTION
**Pull request by Kezii**
The original code was using the wrong overloaded constructor of `rect`, using two points instead of one point and dimension, this patch makes it works like it was originally intended
I added some pixels left and right so it's less tight

before:
immagine

after:
immagine

immagine

immagine

### CODE REVIEW

**Comment by Kezii** [comment]
Wait I'll make it so the border is inside

**Comment by AFCMS** [comment]
Can this be disabled in `minetest.conf`?

**Comment by Kezii** [comment]
> Can this be disabled in `minetest.conf`?

no

**Comment by AFCMS** [comment]
it should

**Comment by AFCMS** [comment]
> he original code was using the wrong overloaded constructor of rect
If the code is a pr, could you link it?

**Comment by Kezii** [comment]
> it should

I don't think this sould go in a config file, if you don't like the proposed implementation, I'd rather ask you for a better solution

the current implementation is bad and needs to change, other than being ugly, I have 10/10 vision and I am not able to read the text comfortably without leaning towards the screen and squinting my eyes

> If the code is a pr, could you link it?

I traced it back to a 6000+ lines diff made more than 10 years ago, given how superficially this was thought out I'm pretty sure this was there since the original implementation and nobody cared to make it decent

**Comment by Kezii** [comment]
immagine
immagine

**Comment by AFCMS** [comment]
better to increase size font

**Comment by wsor4035** [comment]
is this configurable by modders, or is it being forced upon us? also will this affect the PR https://github.com/minetest/minetest/pull/8448 that allows you to modify what is displayed there? tbh i would prefer to see this as a configurable option to modders in addition to the linked PR above

**Comment by Kezii** [comment]
I notice that all of minetest uses the same font size, 

> is this configurable by modders, or is it being forced upon us? also will this affect the PR #8448 that allows you to modify what is displayed there? tbh i would prefer to see this as a configurable option to modders in addition to the linked PR above

I don't see readability as a concern for that PR, also, they seem to have left the wrong constructor bug unchanged.

Since you don't seem to like the dark background, how would you address the fact that the text is almost unreadable on big screens? The font size doesn't seem to scale

I don't see the point of having this configurable by modders, why would a mod want to make the text hard to read?

**Comment by wsor4035** [comment]
> Since you don't seem to like the dark background, how would you address the fact that the text is almost unreadable, especially on big screens? The font size doesn't seem to scale
> 
> I don't see the point of having this configurable by modders, why would a mod want to make the text hard to read?

frankly, I see this like the nametag PR(https://github.com/minetest/minetest/pull/10152), I support it being added as its more readable, however as with that PR, I believe the background color and text color should be configurable to modders, giving them the option to change it to whatever they want for whatever purpose. minetest is a engine and it has to many hard coded things already tbh. see https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L6390-L6408 about setting nametag color and background color. personally for this case I would suggest using meta to do this here

**Comment by Kezii** [comment]
Nah

We are talking about different things, I'm addressing the fact that there is white text on a light background, because it's a stupid default to have: pick my original screenshot, put it full screen, and tell me if you can read the text

It's an accessibility / readability problem that has nothing to do with modding or configuration, if you want the background to be configurable you are free to add that on the next pr, I'm not interested in doing that, I just want to read the text

**Comment by Kezii** [comment]
And again, this is not a "feature", I just fixed the existing code, `driver->draw2DRectangle(bgcolor, rect2, clip);` has been wasting cpu cycles for 10 years

**Comment by wsor4035** [comment]
> Nah
> 
> We are talking about different things, I'm addressing the fact that there is white text on a (potentially) light background, because it's a stupid default to have: pick my original screenshot, put it full screen, and tell me if you can read the text
> 
> It's an accessibility / readability problem that has nothing to do with modding or configuration, if you want the background to be configurable you are free to add that on the next pr, I'm not interested in doing that, I just want to read the text

this can be debated to no end, and yes I can read it fine. However I do believe having the background is better but I still stand by my belief that it should be configurable

**Comment by neoh4x0r** [comment]
> If the code is a pr, could you link it?
> > I traced it back to a 6000+ lines diff made more than 10 years ago, given how superficially this was thought out I'm pretty sure this was there since the original implementation and nobody cared to make it decent

**src/client/hud.cpp** was added in **[CSM] Add basic HUD manipulation. #6067** (about 3 years ago)
commit 9649e4721467dab348011633c814a63a184bd018

> The original code was using the wrong overloaded constructor of rect, using two points instead of one point and dimension

**The original use of rect2** defines the rectangle in the same way that Minetest defines rectangles (by specifying only two points) 
```
v2s32 sdim(dim.X,dim.Y);
core::rect<s32> rect2(
	rect.LowerRightCorner - sdim,
	sdim
);
driver->draw2DRectangle(bgcolor, rect2, clip);
```

**The changes made by this PR**, still only defines the rectangle using two points.
```
const s32 horizontal_padding = 3; //px
core::position2d<s32> offset(-5,-2);

core::rect<s32> background_rect(
	rect.LowerRightCorner - sdim + core::position2d<s32>(-horizontal_padding,0) + offset,
	rect.LowerRightCorner + core::position2d<s32>(horizontal_padding,0) + offset
);
driver->draw2DRectangle(bgcolor, background_rect, clip);
core::rect<s32> rect2(
	rect.LowerRightCorner - sdim + offset,
	rect.LowerRightCorner + offset
);
```

**The rect class is defined in irrlicht** -- the only constructor taking a single point and size is just a helper function for passing in two points.
https://github.com/minetest/irrlicht/blob/master/include/rect.h
```cpp
rect() : UpperLeftCorner(0,0), LowerRightCorner(0,0) {}

rect(T x, T y, T x2, T y2) : UpperLeftCorner(x,y), LowerRightCorner(x2,y2) {}

rect(const position2d<T>& upperLeft, const position2d<T>& lowerRight)
	: UpperLeftCorner(upperLeft), LowerRightCorner(lowerRight) {}

template <class U>
rect(const position2d<T>& pos, const dimension2d<U>& size)
	: UpperLeftCorner(pos), LowerRightCorner(pos.X + size.Width, pos.Y + size.Height) {}

template <class U>
explicit rect(const dimension2d<U>& size)
	: UpperLeftCorner(0,0), LowerRightCorner(size.Width, size.Height) {}
```

**Comment by Kezii** [comment]
> **src/client/hud.cpp** was added in **[CSM] Add basic HUD manipulation. #6067** (about 3 years ago)
> commit [9649e47](https://github.com/minetest/minetest/commit/9649e4721467dab348011633c814a63a184bd018)

No, that pr just moved the file from another path, I know git blame too, read the diff you linked and you'll see that the actual origin of that line of code is well beyond in the past

Also, I don't see how this is relevant



> **The original use of rect2** defines the rectangle in the same way that Minetest defines rectangles (by specifying only two points)
> [...]

No, read the code, in the original code `sdim` is not a point on the screen, even though it has the wrong type, it's the dimension of the text, and it was triggering the wrong constructor of `rect` since two `position2d` were applied, the resulting geometry was wrong and the call to `draw2DRectangle` was useless

You are right about the fact that after my patch there are two points, that is in fact the correct usage of `rect` and what the original author had in mind, and now the rectangle is clearly rendered

**Comment by neoh4x0r** [comment]
@Kezii What inventory mod are you using?

**Comment by hlqkj** [comment]
> TL;DR

Well your post was long too and I read it all...

Besides @sfan5 why has this been labeled as Feature when it's just a fix for a (very old it seems) bug?

I really don't get all this debate over insignificant details. The original code clearly meant to draw a background that never worked cause of a an oversight it seems, let them fix this. Then feature requests can be made for details which are out of the scope of this pr, like letting modders choose the colour, etc. For such things by the way, invlists styling would be best imho.

@neoh4x0r I see no harm on adding a small padding to the text given a rect is now drawn, without that readability would actually be degraded. And same applies for text placement: it should be aligned with the item image like in your last image, not touching the cell bounds, that would again degrade text readability!

**Comment by appgurueu** [comment]
This is a feature because Minetest consistently used to draw text without background rectangles. Only chat has a background box - and only if expanded. The first change that was made to this was Zughy's nametag backgrounds, which got configurability options as well. I don't see why this should not be configurable. This seems to be severely biased towards MineCraft imitation interests.

Just because this changes code to work as intended doesn't mean it now works as expected.

**Comment by hlqkj** [comment]
The chat text has a shadow when not expanded, itemstack count not.

Again, this is **not** a feature and thus it is **not** adding any features, like "configurable itemstack count background color" would be. This is nothing more than a fix of a **bug** that was in the code: just because it compiled it doesn't mean the bug wasn't there. And just because fixing it changes the appearance of the game, doesn't mean its intended to mimic the minecraft design (which by the way, has been literally cloned by minetest so many times, even when it should't have been, that citing this matter here is nonsense IMAO).

And just because no one else noted this bug before doesn't mean there was no bug. The same way, just because no one asked for a background before, doesn't mean that all of us are ok with things like the following (taken from a vanilla MTG server, no mods):

Screenshot 2021-06-29 162725

Not to speak of many other non-MTG items in combination with other inventory mods.

**Comment by appgurueu** [comment]
> The chat text has a shadow when not expanded, itemstack count doesn't.

Plain wrong. Setting `font_shadow = 10` to highlight:

Screenshot


> Again, this is **not** a feature and thus it is **not** adding any features, like "configurable itemstack count background color" would be. This is nothing more than a fix of a **bug** that was in the code: just because it compiled it doesn't mean the bug wasn't there.

> And just because fixing it changes the appearance of the game, doesn't mean its intended to mimic the minecraft design (which by the way, has been literally cloned by minetest so many times, even when it should't have been, that citing this matter here is nonsense

After having a look at a couple more Minecraft screenshots I have to admit that the only thing this got in common with those is it's obtrusiveness.

> And just because no one else noted this bug before doesn't mean there was no bug.

That depends on your definition of bug. At least it wasn't perceived as a bug and rather accepted (or not deemed annoying enough to warrant an issue) and assumed to be a design decision.

> The same way, just because no one asked for a background before, doesn't mean that all of us are ok with things like the following (taken from a vanilla MTG server, no mods):
> 
> Screenshot 2021-06-29 162725

It does heavily imply that it is not much of an issue though.

Let me reiterate: This is an obtrusive design change which doesn't fit in with the rest of the design. Minetest consistently uses font shadows to make text readable everywhere. This PR aims to make font readable using a background box. To the overall issue of text legibility, I see three solutions, which should be configurable and consistent across all text:

* Font shadows: Very unobtrusive, what Minetest currently uses, works fairly well in most cases, a bit problematic for dark text (#9504).
* Font outline: Fairly unobtrusive. Briefly mentioned in the nametag discussion, quite some work to implement. Similar issues as shadows, but slightly better legibility as all borders (not only the bottom right) get an outline.
* Background boxes: The most obtrusive way to improve legibility I can think of. However also the most legible.

**Comment by hlqkj** [comment]
> 
> 
> > The chat text has a shadow when not expanded, itemstack count doesn't.
> 
> Plain wrong. Setting `font_shadow = 10` to highlight:
> 
> Screenshot
> 

My bad, very sorry. I went by observation and didn't noticed it, or at least it didn't helped very much in comparison to this PR (see my image).

> 
> > And just because no one else noted this bug before doesn't mean there was no bug.
> 
> That depends on your definition of bug. At least it wasn't perceived as a bug and rather accepted (or not deemed annoying enough to warrant an issue) and assumed to be a design decision.
> 

Yes indeed, my point stands on a strict definition of bug. I know it also changes the user experience, but since the fixed code was clearly meant to show this bakground I assumed that the thing got in some way already discussed/designed/planned in the past and/or that it wouldn't be such an unbearable ugliness, comparing to the great improvement in readability.

> > The same way, just because no one asked for a background before, doesn't mean that all of us are ok with things like the following (taken from a vanilla MTG server, no mods):
> > Screenshot 2021-06-29 162725
> 
> It does heavily imply that it is not much of an issue though.
> 

Not necessarily. Many of us "users" knows by experience how worth is to propose things here and just give up before even starting: seeing a PR already done I thought of supporting it instead.

> Regarding the "fix vs. feature" discussion: This might be technically considered a fix, but practically, it's a visual _change_, which supposedly - according to you - is an *_improvement_, and therefore a feature. You're just hoping to get this merged with less discussion by misleadingly labelling it a fix based on highly technical reasoning.

A few more words here, since I feel that an undeserved accusation.

Yes I am hoping it will get merged and I see nothing wrong in this. However how could you say I'm hoping to get this passed without discussion when I am spending time here commenting more than its author does, I wonder?!

The labelling thing was just to stress that - in my opinion - this should be considered a bug because the clear intention of drawing the background rect is already in the code, to my eyes implying that someone already designed this in the past but failed to implement it properly. So to keep things strictly on topic, this would be a bug and any other thing about its configurability would be a feature, **and I already stated different times that I'd do agree on having this configurable and themed**. But dropping the whole PR under the disguise that it lacks features just because some doesn't like it it's what concerns me.

There would have been even more chances of a constructive and on-topic discussion if people would discuss how to improve the feature (which is already there) rather than using the "lack of ..." reason to reject this _a-priori_, and labelling this as a feature rather than bug imho started this. This is my point as per the "fix vs. feature" thing.

The author did not introduced anything new in the code, including theming and settings, because it would have been out of the scope of this PR - that is to simply make the background work as it was already intended by someone else. In other words, lack of config and themes was already in the code and this PR **is not** addressing that.

A more clean way to proceed would have been fix the bug, let people test the thing and have a separate discussion to decide if either
- remove it, and we would end up having cleaner code because just rejecting this PR would leave the bug where it is, or
- improve it with everything that would have come out from people tests (maybe more than just a setting to disable and theming).

It's not that a merged PR gets straight into the next release, plenty of time to discuss everything. Unless someone else but me wants to bury this before...

Overall I do agree that this does change user experience and that it is obtrusive, but I also wonder how this would be inconsistent when it has already been put in place for nametags.

**Comment by appgurueu** [comment]
> Yes I am hoping it will get merged and I see nothing wrong in this. However how could you say I'm hoping to get this passed without discussion when I am spending time here commenting more than its author does, I wonder?!

I did not say "without discussion" but "less discussion".

> The labelling thing was just to stress that - in my opinion - this should be considered a bug because the clear intention of drawing the background rect is already in the code, to my eyes implying that someone already designed this in the past but failed to implement it properly.

During testing, they *must* have noticed the absence of the rectangle; had it been important to them, they could have left a comment. Most likely they were satisfied with the result and just left some unneeded code there.

> But dropping the whole PR under the disguise that it lacks features just because some doesn't like it it's what concerns me.

This PR isn't large. Quite the opposite, it's very small. There's no point in splitting up the necessary changes across multiple PRs. I want to prevent this being merged without the configurability being added later on, or even worse, delaying the release.

> Overall I do agree that this does change user experience and that it is obtrusive, but I also wonder how this would be inconsistent when it has already been put in place for nametags.

The nametag thing was quite a controversial mess though. Luckily rubenwardy saved the day in #10937 , but I don't want to rely on that happening again. A feature PR is way harder to get merged than a bugfix, and this burden should not be laid on the next PR to fix the mess this will introduce.

The inconsistency is quite obvious though: All HUD elements miss such a background. Same for other text elements in formspecs. Nametags are the exception, not the norm. Zughy managed to create consistancy anyways, by manually adding background boxes as HUD images behind all HUD text.

**Comment by hlqkj** [comment]
> This PR isn't large. Quite the opposite, it's very small. There's no point in splitting up the necessary changes across multiple PRs. I want to prevent this being merged without the configurability being added later on, or even worse, delaying the release.

Ok, I see what you mean and makes sense. Perhaps a new title would be appropriate then, and a more thoughful discussion rather than approve/reject the PR itself.

> The inconsistency is quite obvious though: All HUD elements miss such a background. Same for other text elements in formspecs. Nametags are the exception, not the norm. Zughy managed to create consistancy anyways, by manually adding background boxes as HUD images behind all HUD text.

Also agree on consistency, but since Zughy's (and rubenwardy's) changes got merged would actually be a consistency improvement, to slowly migrate everything else towards that direction. That is, having the background as an option to the player and let modders style it at their wish.

**Comment by neoh4x0r** [comment]
>>appgurueu 
>This PR isn't large. Quite the opposite, it's very small. There's no point in splitting up the necessary changes across multiple >PRs. I want to prevent this being merged without the configurability being added later on, or even worse, delaying the release.

We don't really need to compound [this] bugfix with unrelated changes. 
If someone really want's those features they should open a new PR (and mention this one).
Cramming too much stuff into this PR is likely cause its merge to be stalled/delayed.

>>hlqkj 

> The author did not introduced anything new in the code

The trouble is that this PR is a mix of a bug fix and feature request.

The "*bug fix*" was **to make the background rect work behind the item count** (changing the definition of sdim would be enough to do this)

The feature was to **add an offset so that the item count/background are more inset from the frame** (a purely cosmetic fix, since it is not correcting a mistake in the code).

Not to mention the "feature creep" of making it configurable, etc -- which is beyond a mere bug-fix.


As I have already mentioned in a previous comment the added offset get exaggerated and is linked to various things (such as font size, resolution, etc, etc).

If we want to make things consistent -- then this would be the antithesis to that endeavor (it would need to look the same in every situation). 

Another point of contention is whether we could actually classify this as a bug.

[https://en.wikipedia.org/wiki/Software_bug](https://en.wikipedia.org/wiki/Software_bug)
> A software bug is an error, flaw or fault in a computer program or system that causes it to produce an incorrect or unexpected result, or to behave in unintended ways.

Technically, the code was not using the correct constructor -- as @Kezii has pointed out.

However, the "bug" has been there for so long it might not be perceived (by users) as being "incorrect, unexpected, or behaving unintendedly".

Moreover, this this "bug" is cosmetic/visual (meaning it should probably have low priority) -- it is not a grave game-breaking bug (it just a cosmetic fix).

Furthermore, I don't think it is a very big deal because no one has reported (or cared enough to report) this issue, until now.

One thing that is very apparent is the screenshots provided in the first post cause the bug/issue to standout more because of the light background in the inventory form (**which seems to be implementation of minecraft inventory**); which might be more related to how the mod draw the form than it is with minetest.

Also MTG's inventory form (with either the default or unified inventory installed) looks fine without this fix.

Which leads me to question whether users will have very different experiences based on how their game settings are configured (and what they use for an inventory mod) -- vis-a-vis this "bugfix" might actually cause more issues than it fixes.

**Comment by hlqkj** [comment]
> The feature was to **add an offset so that the item count/background are more inset from the frame** (a purely cosmetic fix, since it is not correcting a mistake in the code).

True, however as I tried to remark, this is a needed change for the background to be an effective improvement over the plain text.

> As I have already mentioned in a previous comment the added offset get exaggerated and is linked to various things (such as font size, resolution, etc, etc).

And it should be linked to those things. I've got tons of mods which I had to fix myself that had their formspecs destroyed by the DPI/font scaling once it got introduced (I think it was with 5.4 IIRC?). The font scaling is a system setting and minetest, as other apps/programs now respect it: you wouldn't like a larger font with the background rect sized for the default one, no?

> One thing that is very apparent is the screenshots provided in the first post cause the bug/issue to standout more because of the light background in the inventory form (**which seems to be implementation of minecraft inventory**); which might be more related to how the mod draw the form than it is with minetest.

True, but then let game makers have the option to use light colors without making players lives too difficult, I say.

> Also MTG's inventory form (with either the default or unified inventory installed) looks fine without this fix.

Also depends on the item image, see below.

> Screenshot 2021-06-29 162725

**Comment by neoh4x0r** [comment]
> hlqkj
> Also depends on the item image, see below.
> 
> Screenshot 2021-06-29 162725

Depending on the font size, it is possible that the added offset/inset will end up covering up too much of the inventory image.
Also the itemcount shown on the glass pane in your screenshot seems to be pushed up more from the bottom of the inventory frame's bottom edge then mine is (on MTG with default inventory mod).

minetest-issue-11316

**IMHO: the itemcount should look like this**

minetest-issue-11316-expected

**Comment by hlqkj** [comment]
What about a compromise, that is background bottom-right aligned but having a (very small) padding between text and background?

Also, try this with the glass fragments item, not the glass pane: that was actually there only to show the contrast between a fairly readable one (pane) and one almost completely impossible to read (glass fragments).

**Comment by EliasFleckenstein03** [comment]
Please make this configurable.

**Comment by Df458** [comment]
There was some discussion of this in today's meeting, here is my take:
> Personally I don't like the rectangle at all, but I see the value for certain games or users with vision disabilities--it seems to me like the easiest way to get it in is as an accessibility setting, then expose more to modders if anyone actually requests it later
> My *suspicion* is that people asking for the ability for games to configure it actually just want the option not to have it

I think this is a genuinely useful accessibility feature, but I also understand why some players wouldn't want to have it forced on them. I also think arguing over semantics here is a waste of everyone's time.

So, I propose that you add a setting in the client for this. Could easily have a default background opacity setting going from 0 to 1, and I think this will cover the vast majority of players' needs. If modders want this to then be configurable in mods, we can discuss a possible follow-up PR. But I don't see that as necessary for this PR.

**Comment by Zughy** [comment]
@Df458 I don't think this should be optional (we have plenty of pretty useless optional settings already), just find a balance between readability and aesthetic to offer saner default. My two cents:
- black outline around the text (quick mock-up: https://imgur.com/a/pDrUhc8)
- increase text size (see https://github.com/minetest/minetest/pull/11316#issuecomment-855253502)
- default font is dull, but that's for another day

I can't see a case where people say "no, let me remove the outline, I want this text to be less readable"
